### DISCUSSION TITLE
API response refactor/suggestions

### DISCUSSION THREAD

**Discussion by enrichman** [epinio/epinio#1188] [Ideas]
## Empty responses

Playing around with the APIs I've seen some response are not always returning a JSON, i.e. if you are not authorized then an empty body is returned:

```
> GET /api/v1/namespaces HTTP/1.1
> Host: localhost:8080
> Authorization: Basic YWRtaW46ajNxcVlDQ2FkWkZxXA==
> User-Agent: curl/7.81.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 401 Unauthorized
* Authentication problem. Ignoring this.
< Www-Authenticate: Basic realm="Authorization Required"
< Date: Fri, 04 Feb 2022 11:48:10 GMT
< Content-Length: 0
< 
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
* Connection #0 to host localhost left intact
```

This could cause some issues or problem with the unmarshalling of the response on client side, since you cannot rely on the status code or the body. You should do some error checking on the unmarshalling.

### proposal

I would suggest to always return a JSON, and if the response is NOT in the range of the success (200/299, or redirects 300/399) but is >=400 then our API error is returned:

```json
{
  "errors": [
    {
      "status": 500,
      "title": "something bad",
      "details": "error details"
    }
  ]
}
```

## Errors

I've seen also that when the server is failing sometimes is throwing out the whole stacktrace. Usually you don't want to expose internal details, so I suggest to hide this information.

```json
{
  "errors": [
    {
      "status": 500,
      "title": "something bad",
      "details": "\nServer Backtrace: something bad\ngithub.com/epinio/epinio/internal/namespaces.List\n\t/home/enrico/Development/SUSE/epinio/epinio/internal/namespaces/namespaces.go:27\ngithub.com/epinio/epinio/internal/api/v1/namespace.Controller.Index\n\t/home/enrico/Development/SUSE/epinio/epinio/internal/api/v1/namespace/index.go:29\ngithub.com/epinio/epinio/internal/api/v1.errorHandler.func1\n\t/home/enrico/Development/SUSE/epinio/epinio/internal/api/v1/router.go:40\ngithub.com/gin-gonic/gin.(*Context).Next\n\t/home/enrico/.asdf/installs/golang/1.17.6/packages/pkg/mod/github.com/gin-gonic/gin@v1.7.7/context.go:168\ngithub.com/gin-gonic/gin.LoggerWithConfig.func1\n\t/home/enrico/.asdf/installs/golang/1.17.6/packages/pkg/mod/github.com/gin-gonic/gin@v1.7.7/logger.go:241\ngithub.com/gin-gonic/gin.(*Context).Next\n\t/home/enrico/.asdf/installs/golang/1.17.6/packages/pkg/mod/github.com/gin-gonic/gin@v1.7.7/context.go:168\ngithub.com/gin-contrib/sessions.Sessions.func1\n\t/home/enrico/.asdf/installs/golang/1.17.6/packages/pkg/mod/github.com/gin-contrib/sessions@v0.0.4/sessions.go:54\ngithub.com/gin-gonic/gin.(*Context).Next\n\t/home/enrico/.asdf/installs/golang/1.17.6/packages/pkg/mod/github.com/gin-gonic/gin@v1.7.7/context.go:168\ngithub.com/gin-gonic/gin.CustomRecoveryWithWriter.func1\n\t/home/enrico/.asdf/installs/golang/1.17.6/packages/pkg/mod/github.com/gin-gonic/gin@v1.7.7/recovery.go:99\ngithub.com/gin-gonic/gin.(*Context).Next\n\t/home/enrico/.asdf/installs/golang/1.17.6/packages/pkg/mod/github.com/gin-gonic/gin@v1.7.7/context.go:168\ngithub.com/gin-gonic/gin.(*Engine).handleHTTPRequest\n\t/home/enrico/.asdf/installs/golang/1.17.6/packages/pkg/mod/github.com/gin-gonic/gin@v1.7.7/gin.go:555\ngithub.com/gin-gonic/gin.(*Engine).ServeHTTP\n\t/home/enrico/.asdf/installs/golang/1.17.6/packages/pkg/mod/github.com/gin-gonic/gin@v1.7.7/gin.go:511\nnet/http.serverHandler.ServeHTTP\n\t/home/enrico/.asdf/installs/golang/1.17.6/go/src/net/http/server.go:2879\nnet/http.(*conn).serve\n\t/home/enrico/.asdf/installs/golang/1.17.6/go/src/net/http/server.go:1930\nruntime.goexit\n\t/home/enrico/.asdf/installs/golang/1.17.6/go/src/runtime/asm_amd64.s:1581"
    }
  ]
}

```

### Empty logs

Also this information is NOT logged on the server side, that is much more an issue, since logs without details are not really useful.

```
Method: GET | Path: /api/v1/namespaces | User: admin | IP: 127.0.0.1 | Timestamp: 2022-02-04T12:49:50+01:00 | Latency: 13.652588ms | Status: 500 | Message:
```

**Comment by andreas-kupries** [comment]
The big backtrace is the `%+v` on the error value at

https://github.com/epinio/epinio/blob/main/pkg/api/core/v1/errors/errors.go#L86

**Reply by manno** [reply]
Yes, however I learned, that stacktraces only work for errors generated by https://github.com/pkg/errors#roadmap, which is now in maintenance mode. Ah and the manually created ones, too.

**Comment by andreas-kupries** [comment]
Logging of internal errors could possibly be done via a gin-gonic middle-ware running after the regular handler !?

**Reply by enrichman** [reply]
Yes, a centralized logging middleware would be the best option!

**Comment by manno** [comment]
* we could have a 'development'  mode where errors do return the stacktrace, but you're right these belong in the logs first and foremost.
* we used to have Traefik handle basic auth, but now we have Gin middleware for that. I agree we should add a JSON body to all our auth errors to JSON. However that should not have too many details for security reasons. Also, one could argue that the status code is enough for most clients?

**Reply by enrichman** [reply]
+1 for the development mode, that would be interesting.

It could be enough for a 401 but maybe not for a 403, and not for the 400 bad request, so it would be "complicated" to write a centralized unmarshaller for the responses:

```
if code is 401 {
    return unauthenticated error (without unmarshal)
}
if code is 403 or 400 {
   // maybe we have an error with some details about the unauth, like "not auth for this workspace", so you should unmarshal this
    unmarhsalRespErr(resp)
}
else if success {
    unmarshalResp(resp)
}
```

so it's just a matter of simplicity. Maybe you can even have a single ApiResponse struct that can handle both success/error cases.

We can probably then simplify this part: https://github.com/epinio/epinio/blob/6663725dbeb5b3ecb06d0daf17135b0b3b2845d6/pkg/api/core/v1/client/http.go#L179

**Reply by manno** [reply]
That func is only used in one place I think. We should already be able to refactor that and delete that func in favor of `do()`. 
This is the generic error response we use now: https://github.com/epinio/epinio/blob/main/pkg/api/core/v1/errors/errors.go#L9-L12

We also have this https://github.com/epinio/epinio/blob/6663725dbeb5b3ecb06d0daf17135b0b3b2845d6/internal/api/v1/response/response.go#L41-L51
